actors/builtin/power/power_test.go

```diff
@@ -72,6 +72,243 @@ func TestConstruction(t *testing.T) {
+			ConstructorParams: initCreateMinerBytes(t, owner, owner, peer, mAddr, sealProofType),
+		}
+		rt.ExpectSend(builtin.InitActorAddr, builtin.MethodsInit.Exec, msgParams, abi.NewTokenAmount(10),
+			nil, exitcode.ErrInsufficientFunds)
+
+		rt.ExpectAbort(exitcode.ErrInsufficientFunds, func() {
+			rt.Call(ac.CreateMiner, createMinerParams)
+		})
+	})
+}
+
+func TestUpdateClaimedPowerFailures(t *testing.T) {
+	rawDelta := big.NewInt(100)
+	qaDelta := big.NewInt(200)
+	miner := tutil.NewIDAddr(t, 101)
+
+	t.Run("fails if caller is not a StorageMinerActor", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		params := power.UpdateClaimedPowerParams{
+			RawByteDelta:         rawDelta,
+			QualityAdjustedDelta: qaDelta,
+		}
+		rt.SetCaller(miner, builtin.SystemActorCodeID)
+		rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+
+		rt.ExpectAbort(exitcode.ErrForbidden, func() {
+			rt.Call(ac.UpdateClaimedPower, &params)
+		})
+
+		rt.Verify()
+	})
+
+	t.Run("fails if claim does not exist for caller", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		params := power.UpdateClaimedPowerParams{
+			RawByteDelta:         rawDelta,
+			QualityAdjustedDelta: qaDelta,
+		}
+		rt.SetCaller(miner, builtin.StorageMinerActorCodeID)
+		rt.ExpectValidateCallerType(builtin.StorageMinerActorCodeID)
+
+		rt.ExpectAbort(exitcode.ErrIllegalState, func() {
+			rt.Call(ac.UpdateClaimedPower, &params)
+		})
+
+		rt.Verify()
+	})
+}
+
+func TestEnrollCronEpoch(t *testing.T) {
+	miner := tutil.NewIDAddr(t, 101)
+
+	t.Run("enroll multiple events", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		e1 := abi.ChainEpoch(1)
+
+		// enroll event with miner 1
+		p1 := []byte("hello")
+		ac.enrollCronEvent(rt, miner, e1, p1)
+
+		events := ac.getEnrolledCronTicks(rt, e1)
+
+		var evt power.CronEvent
+
+		found, err := events.Get(0, &evt)
+		require.True(t, found)
+		require.NoError(t, err)
+		require.EqualValues(t, p1, evt.CallbackPayload)
+		require.EqualValues(t, miner, evt.MinerAddr)
+
+		// enroll another event with the same miner
+		p2 := []byte("hello2")
+		ac.enrollCronEvent(rt, miner, e1, p2)
+		events = ac.getEnrolledCronTicks(rt, e1)
+		found, err = events.Get(0, &evt)
+		require.True(t, found)
+		require.NoError(t, err)
+		require.EqualValues(t, p1, evt.CallbackPayload)
+		require.EqualValues(t, miner, evt.MinerAddr)
+
+		found, err = events.Get(1, &evt)
+		require.NoError(t, err)
+		require.True(t, found)
+		require.EqualValues(t, p2, evt.CallbackPayload)
+		require.EqualValues(t, miner, evt.MinerAddr)
+
+		// enroll another event with a different miner for a different epoch
+		e2 := abi.ChainEpoch(2)
+		p3 := []byte("test")
+		miner2 := tutil.NewIDAddr(t, 501)
+		ac.enrollCronEvent(rt, miner2, e2, p3)
+		events = ac.getEnrolledCronTicks(rt, e2)
+		found, err = events.Get(0, &evt)
+		require.NoError(t, err)
+		require.True(t, found)
+		require.EqualValues(t, p3, evt.CallbackPayload)
+		require.EqualValues(t, miner2, evt.MinerAddr)
+	})
```

Comment:
Please test enrolling for an epoch that is before the current epoch

Reply:
This is done.